nibabel/streamlines/tractogram.py

```diff
@@ -0,0 +1,543 @@
+            :class:`LazyTractogram` object is returned. Otherwise, streamlines
+            are modified in-place.
+
+        Returns
+        -------
+        tractogram : :class:`Tractogram` or :class:`LazyTractogram` object
+            Tractogram where the streamlines have been transformed according
+            to the given affine transformation. If the `lazy` option is true,
+            it returns a :class:`LazyTractogram` object, otherwise it returns a
+            reference to this :class:`Tractogram` object with updated
+            streamlines.
+        """
+        if lazy:
+            lazy_tractogram = LazyTractogram.from_tractogram(self)
+            lazy_tractogram.apply_affine(affine)
+            return lazy_tractogram
+
+        if len(self.streamlines) == 0:
+            return self
+
+        BUFFER_SIZE = 10000000  # About 128 Mb since pts shape is 3.
+        for i in range(0, len(self.streamlines._data), BUFFER_SIZE):
+            pts = self.streamlines._data[i:i+BUFFER_SIZE]
+            self.streamlines._data[i:i+BUFFER_SIZE] = apply_affine(affine, pts)
+
+        # Update the affine that brings back the streamlines to RASmm.
+        self._affine_to_rasmm = np.dot(self._affine_to_rasmm,
+                                       np.linalg.inv(affine))
+
+        return self
+
+
+class LazyTractogram(Tractogram):
+    """ Class containing information about streamlines.
+
+    Tractogram objects have four main properties: `header`, `streamlines`,
+    `scalars` and `properties`. Tractogram objects are iterable and
+    produce tuple of `streamlines`, `scalars` and `properties` for each
+    streamline.
+
+    Notes
+    -----
+    If provided, `scalars` and `properties` must yield the same number of
+    values as `streamlines`.
+    """
+
+    class LazyDict(collections.MutableMapping):
+        """ Internal dictionary with lazy evaluations. """
+
+        def __init__(self, *args, **kwargs):
+            self.store = dict()
+
+            # Use update to set keys.
+            if len(args) == 1 and isinstance(args[0], LazyTractogram.LazyDict):
+                self.update(dict(args[0].store.items()))
+            else:
+                self.update(dict(*args, **kwargs))
+
+        def __getitem__(self, key):
+            return self.store[key]()
+
+        def __setitem__(self, key, value):
+            if value is not None and not callable(value):
+                raise TypeError("`value` must be a coroutine or None.")
+
+            self.store[key] = value
+
+        def __delitem__(self, key):
+            del self.store[key]
+
+        def __iter__(self):
+            return iter(self.store)
+
+        def __len__(self):
+            return len(self.store)
+
+    def __init__(self, streamlines=None,
+                 data_per_streamline=None,
+                 data_per_point=None):
+        """
+        Parameters
+        ----------
+        streamlines : coroutine yielding ndarrays of shape (Nt,3) (optional)
```

Comment:
I agree with you. I used coroutine because I was lacking a better term. What should be passed to a `LazyTractogram` is a "**function that returns** an instantiated generator" so it can be called multiple times. 

Okay, I don't know why I didn't find the term earlier but it seems those are simply called "generator function" ;) https://wiki.python.org/moin/Generators. I'll replace coroutine with that. Sounds good?